hail/src/main/scala/is/hail/expr/Parser.scala

```diff
@@ -9,22 +9,21 @@ import is.hail.table.{Ascending, Descending, SortField}
 import is.hail.utils.StringEscapeUtils._
 import is.hail.utils._
 import is.hail.variant._
-import org.apache.spark.sql.Row
 import org.json4s.jackson.{JsonMethods, Serialization}
 
 import scala.util.parsing.combinator.JavaTokenParsers
 import scala.util.parsing.input.{Position, Positional}
+import scala.collection.JavaConverters._
 
 case class Positioned[T](x: T) extends Positional
 
-object IRParserEnvironment {
-  var theEnv: IRParserEnvironment = null
-}
-
 case class IRParserEnvironment(
-  ref_map: Map[String, Type] = Map.empty,
-  ir_map: Map[String, BaseIR] = Map.empty
-)
+  refMap: Map[String, Type] = Map.empty,
+  irMap: Map[String, BaseIR] = Map.empty
+) {
+  def update(newRefMap: Map[String, Type] = Map.empty, newIRMap: Map[String, BaseIR] = Map.empty): IRParserEnvironment =
```

Comment:
Few things:
 - in relational IR scope, the refMap should be empty,
 - in practice you don't update newIRMap,
 - so I think there should be `.withRefMap` for the relational case, that should assert the refMap is empty, and
 - def +(t: (String, Type)) to add a single binding in for Let.

Reply:
I'm concerned about the case where it's something like this and the reference defined in the outer scope is lost if we assert it should be empty:
Let("x", MatrixRead(...), MatrixUnionRows(MatrixRead(...) Ref("x")))

Reply:
`Let` can only be used for values currently, not relational IR.

Reply:
I guess you can do `Let(x, 5, Ref(x) + TableAggregate(t, ...))`, in which case the refMap wouldn't be empty.  But for this, x isn't visible in the TableAggregate, so you should pass in an empty refMap.

Reply:
Which is to say, I think what you did is correct.